Update handling of opacity values to match latest spec resolution

weinig opened the pull request:
#### ab6e87bdf57f58aaef6a3eda1f7f7e02557230b4
<pre>
Update handling of opacity values to match latest spec resolution
<a href="https://bugs.webkit.org/show_bug.cgi?id=296725">https://bugs.webkit.org/show_bug.cgi?id=296725</a>

Reviewed by Darin Adler.

Updates the parsing and resolution of opacity values to match latest spec
resolution in w3c/csswg-drafts#10426

Opacity values now parse using the `consumePercentageDividedBy100OrNumber`
rules, meaning that percentage tokens get converted to numbers at parse
time, but calc() functions are not converted until style building. This
means updating both the CSSProperties.json grammars and the fast path.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-valid.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:

Canonical link: <a href="https://commits.webkit.org/298099@main">https://commits.webkit.org/298099@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e886310fdaac8bf5e9034919d58bc79c3855c9a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33969 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120388 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64958 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116111 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42531 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86801 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117170 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27551 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102586 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67190 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26733 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20712 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64083 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20829 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123605 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41241 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30749 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95632 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41618 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98788 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95415 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24320 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40558 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18385 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41120 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46631 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40727 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44033 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42478 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -162,25 +162,25 @@ static inline bool parseSimpleAngle(std::span<const CharacterType> characters, R
-    unit = CSSUnitType::CSS_NUMBER;
+    bool isPercentage = false;
     if (!characters.empty() && characters.back() == '%') {
         dropLast(characters);
```
Is it desirable to do this (without undoing it) even if we fail to parse a number and subsequently return `false`?

weinig replied:
Generally, I would say no, as I prefer these kind of consume functions to do nothing if they fail, it doesn't seem like that is followed in the fast path, so for now, I am think keeping it as is makes sense.

darinadler commented on Source/WebCore/style/values/color/StyleOpacity.h:
```diff
@@ -0,0 +1,73 @@
+    constexpr bool isZero() const { return value == 0; }
+
+    constexpr bool isFullyTransparent() const { return value == 0; }
+    constexpr bool isFullyOpaque() const { return value == 1; }
```
I would have called these functions `isTransparent` and `isOpaque` and I think we use those names for similar functions on colors with alpha. I understand that these longer names are less ambiguous and possibly clearer, so I won't insist on changing them.

darinadler reviewed (approved):
Is there an easy way to make the behavior change in a smaller subsequent separate change or would it be a headache to do the refactoring without correcting the behavior?

weinig committed:
Update handling of opacity values to match latest spec resolution
https://bugs.webkit.org/show_bug.cgi?id=296725

Reviewed by Darin Adler.

Updates the parsing and resolution of opacity values to match latest spec
resolution in w3c/csswg-drafts#10426

Opacity values now parse using the `consumePercentageDividedBy100OrNumber`
rules, meaning that percentage tokens get converted to numbers at parse
time, but calc() functions are not converted until style building. This
means updating both the CSSProperties.json grammars and the fast path.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/opacity-valid.html:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:

Canonical link: https://commits.webkit.org/298099@main
